### PR TITLE
TYCK: add inheritUniv() api

### PR DESCRIPTION
**Pull request by tsao-chi**
This is a part of https://github.com/aya-prover/aya-dev/issues/448

### CODE REVIEW

**Review by ice1000 — COMMENTED** [review]
+ `inheritUniv` -> `universe` (and `doUniverse`), and if you hate this exaggerated name you can rename it everywhere to `Sort` I guess (just make sure the naming is consistent, I don't mind huge rename commits, but please submit standalone PRs for renames)
+ It doesn't have to return a `Result`. I think we only need a `Term`. I think the same applies to `inherit`, but I'm unsure why we're keeping it. Maybe it's time to change.
+ Rest lgtm

**Review by ice1000 — COMMENTED** [review]
+ inheritUniv -> universe (and doUniverse), and if you hate this exaggerated name you can rename it everywhere to Sort I guess (just make sure the naming is consistent, I don't mind huge rename commits, but please submit standalone PRs for renames)
+ It doesn't have to return a Result. I think we only need a Term. I think the same applies to inherit, but I'm unsure why we're keeping it. Maybe it's time to change.
+ universe should take an optional upper bound. Since this is a parameter and is not passed to the constructor of FormTerm.Univ it is okay to use -1 this time.
+ Rest lgtm

**Comment by tsao-chi** [comment]
> It doesn't have to return a Result. I think we only need a Term. I think the same applies to inherit, but I'm unsure why we're keeping it. Maybe it's time to change.

I think Result might be necessary because these 2 functions do process type information

**Comment by ice1000** [comment]
@tsao-chi You're right.

What about using -1? I think it should be more memory efficient.

File: base/src/main/java/org/aya/tyck/ExprTycker.java [hunk 1]
```diff
@@ -497,6 +497,33 @@ yield switch (term.normalize(state, NormalizeMode.WHNF)) {
     };
   }
 
+  private @NotNull Result doUniverse(@NotNull Expr expr, @NotNull Option<Integer> upperBound) {
+    if (upperBound.isDefined()) assert (upperBound.get() >= 0);
+    Term univ = FormTerm.Univ.ZERO;
+    return switch (expr) {
+      case Expr.TupExpr tuple -> fail(tuple, univ, BadTypeError.sigmaCon(state, tuple, univ));
+      case Expr.HoleExpr hole -> {
+        // TODO[lift-meta]: should be able to solve a lifted meta
```
Comment (ice1000) [inline base/src/main/java/org/aya/tyck/ExprTycker.java:506]: We no longer need it

**Review by ice1000 — COMMENTED** [review]
Can we use `universe` in synth and inherit?

**Comment by tsao-chi** [comment]
@tsao-chi

**Review by ice1000 — COMMENTED** [review]
Did you implement `doUniverse(Pi)` and `doUniverse(Sigma)` and `doUniverse(DataCall)? Those are needed

**Comment by tsao-chi** [comment]
> Did you implement `doUniverse(Pi)` and `doUniverse(Sigma)` and `doUniverse(DataCall)? Those are needed

No, synthesize() are re-used for those terms

**Review by ice1000 — COMMENTED** [review]
I'm talking about a function `f(u)` that checks if `u` is a valid type, so apart from universes, you also move the logic of checking `Pi`/`Sigma` into the new function as well. Also the return type semantics needs to be changed: it's the level of the type of the type, not the type itself.

**Comment by ice1000** [comment]
> > Did you implement `doUniverse(Pi)` and `doUniverse(Sigma)` and `doUniverse(DataCall)? Those are needed
> 
> No, synthesize() are re-used for those terms

Can you put everything together? Because synthesize don't work for `Pi`. In fact, `Pi` is overloaded:

+ `A -> B : Type` for `A, B : Type` or `A : ISet`
+ `A -> B : ISet` for `A, B : ISet`
+ `A -> B : Set` for `A, B : Set`

We need `universe` to handle all these

**Comment by ice1000** [comment]
Note that `Set` and `Type` are both universe polymorphic

**Comment by ice1000** [comment]
Gooooooooooooooooooooooood

**Pull request merged**